lib/ansible/module_utils/basic.py

```diff
@@ -1848,9 +1848,9 @@ def safe_eval(self, value, locals=None, include_exceptions=False):
     def _check_type_str(self, value):
         if isinstance(value, string_types):
             return value
-        # Note: This could throw a unicode error if value's __str__() method
-        # returns non-ascii.  Have to port utils.to_bytes() if that happens
-        return str(value)
+        if isinstance(value, (list, dict)):
+            self.warn("Parameter value %s is expected to be string but was %s" % (value, value.__class__.__name__))
+        return to_text(value)
```

Comment:
@willthames bcoca asked me to look at your problem.  I think that we want to use to_native() here and we probably want to use errors='surrogate_or_strict' as well.  to_native() converts the string into the type which an undecorated string literal has on that version of python.  That type is called "str" on both python2 and python3 but it's a byte string on python2 and a text string on python3.  It will have the following benefits:

* It's the most compatible since it's the same type as str on both platforms.  (It just has better defaults  and more flexible error handling than using str()).
* For most module_utils code, we use native string types right now so that a module author can mostly write idiomatic python and it mostly just works.  They don't run into a problem combining byte and text type.  (They do have to worry about it when they deal with an external API which needs a specific type but those are not as common so it's easier to deal with those when the time comes and not worry the rest of the time).

Reply:
@abadger that just leaves the question of what to do with the test - it's failing because the result of `to_text` isn't an instance of type `_type`, which defaults to `str`.